src/network/socket.cpp

```diff
@@ -0,0 +1,453 @@
+bool Address::operator == (const Address &other)
+{
+	if (other.family != family || other.port != port)
+		return false;
+
+	switch (family) {
+	case AF_INET: return host.v4.s_addr == other.host.v4.s_addr;
+	case AF_INET6: return memcmp(host.v6.s6_addr,
+			other.host.v6.s6_addr, 16) == 0;
+	default:
+		return false;
+	}
+}
+
+void Address::resolve(const std::string &name)
+{
+	if (name.empty())
+		return setZero();
+
+	struct addrinfo *resolved, hints = {0};
+
+	// Setup hints
+	// AF_UNSPEC allows both IPv6 and IPv4 addresses to be returned
+	hints.ai_family = g_settings->getBool("enable_ipv6") ?
+			AF_UNSPEC : AF_INET;
+
+	// Do getaddrinfo()
+	int err = getaddrinfo(name.c_str(), NULL, &hints, &resolved);
+	if (err)
+		throw ResolveError(gai_strerror(err));
+
+	// Copy data
+	switch (resolved->ai_family) {
+	case AF_INET:
+		host.v4 = reinterpret_cast<sockaddr_in *>(resolved->ai_addr)->sin_addr;
+		break;
+	case AF_INET6:
+		host.v6 = reinterpret_cast<sockaddr_in6*>(resolved->ai_addr)->sin6_addr;
+		break;
+	default:
+		freeaddrinfo(resolved);
+		throw ResolveError("Unknown address family");
+	}
+	family = resolved->ai_family;
+	freeaddrinfo(resolved);
+}
+
+void Address::deserialize(const std::string &str)
+{
+	std::istringstream is(str);
+	bool bracketed = (is.peek() == '[');
+
+	if (bracketed)
+		is.ignore();
+
+	std::string addr;
+	std::getline(is, addr, bracketed ? ']' : ':');
+	resolve(addr);
+
+	if (bracketed)
+		is.ignore();
+
+	is >> port;
+}
+
+std::string Address::serializeHost() const
+{
+	// Windows XP doesn't have inet_ntop, so use getnameinfo
+	char str[INET6_ADDRSTRLEN];
+	SockAddrUnion addr;
+	int addr_sz = getSocketAddress(&addr);
+	if (int err = getnameinfo(reinterpret_cast<const sockaddr *>(&(addr)),
+			addr_sz, str, INET6_ADDRSTRLEN,
+			NULL, 0, NI_NUMERICHOST))
+		throw SerializationError(
+			std::string("Unable to serialize address: ") +
+				gai_strerror(err));
+	return str;
+}
+
+void Address::serialize(std::ostream &os) const
+{
+	if (isIPv6())
+		os << '[';
+	os << serializeHost();
+	if (isIPv6())
+		os << ']';
+	os << ':' << std::dec << port;
+}
+
+std::string Address::serialize() const
+{
+	std::ostringstream os;
+	serialize(os);
+	return os.str();
+}
+
+bool Address::isZero() const
+{
+	switch (family) {
+	case AF_INET: return host.v4.s_addr == 0;
+	case AF_INET6: return memcmp(&host.v6, &in6addr_any, 16) == 0;
+	}
+	return false;
+}
+
+socklen_t Address::getSocketAddress(SockAddrUnion *addr) const
+{
+	memset(addr, 0, sizeof(*addr));
+	switch (family) {
+	case AF_INET:
+		addr->v4.sin_family = family;
+		addr->v4.sin_port = htons(port);
+		addr->v4.sin_addr = host.v4;
+		return sizeof(addr->v4);
+	case AF_INET6:
+		addr->v6.sin6_family = family;
+		addr->v6.sin6_port = htons(port);
+		addr->v6.sin6_addr = host.v6;
+		return sizeof(addr->v6);
+	}
+	return 0;
+}
+
+bool Address::isLoopback() const
+{
+	switch (family) {
+	case AF_INET: return ntohl(host.v4.s_addr) == INADDR_LOOPBACK;
```

Comment:
AFAIK the whole `127.0.0.0/8` network is defined as loopback. `INADDR_LOOPBACK` is just a single address in it (`127.0.0.1`).
Note: `127.0.0.2` works perfectly on Linux. Not sure about Windows, it may have some problems with that.